Add Appengine Flexible documentation with support for Datastore (#227)
Pylons/pyramid_cookbook: docs/deployment/gae_flex_datastore.rst

.. _appengine_flexible_tutorial:

Google App Engine Flexible with Datastore and :app:`Pyramid`
============================================================

> It is possible to run a :app:`Pyramid` application on `Google App Engine <https://cloud.google.com/appengine/>`_.  This tutorial is written "environement agnostic" meaning the commands here should work on Linux, Mac or Windows. This tutorial also assumes you've already installed and created a :app:`Pyramid` application, and that you have a Google App Engine account.

Reviewer: - Please use one line per sentence for better diffs.
- s/"environement agnostic" meaning/"environment agnostic", meaning
- s/Mac or/macOS, or

Setup
-----

First we'll need to set up a few things in App Engine. If you don't need Datastore access for your project or any other GCP service, you can skip the Credentials section.

Credentials
###########

Navigate to App Engine's IAM And Admin section and click on Service Accounts in the left sidebar, then create a `Service Account <https://cloud.google.com/iam/docs/service-accounts>`_.

Once a service account is created, you will be given a ``.json`` key file. This will be used to allow your Pyramid application to communicate with GCP services. Move this file to your Pyramid project. A best practice here would be to make sure this file is listed in ``.gitignore`` so that it's not checked in with the rest of your code.

> Now that we have a service account, we'll need to give it a couple of roles. Click *IAM* in the left sidebar of *IAM And Admin*. Find the service account you've just created and click the Edit button. Give this account the *Cloud Datastore User* role for read/write access. For read-only access, give it *Cloud Datastore Viewer*.

Reviewer: All GUI interface items should use `:guilabel:`, for example:

```
Click :guilabel:`IAM` in
```

For this paragraph, please do so for `IAM`, `IAM And Admin`, and `Edit`.

I am not sure whether "Cloud Datastore User" and "Cloud Datastore Viewer" are UI elements or literals. If they are literals, then they should use inline rst syntax like so:

```
``Cloud Datastore Viewer``
```

Author: I suppose they are both, technically? https://i.imgur.com/KFy6qnu.png

Reviewer: It looks like a UI element to me, like an option in a select menu that can be clicked, so `:guilabel:` is the way to go. Inline literals are usually for stuff like "enter the command `wherecani get-a-haircut`"

#.  ``dockerfile``

    .. code-block:: docker

        FROM gcr.io/google-appengine/python
        # Create a virtualenv for dependencies. This isolates these packages from
        # system-level packages.
        # Use -p python3 or -p python3.7 to select python version. Default is version 2.
        RUN virtualenv /env -p python3

        # Setting these environment variables are the same as running
        # source /env/bin/activate.
        ENV VIRTUAL_ENV /env
        ENV PATH /env/bin:$PATH
        ENV PYTHONUNBUFFERED 0

        # Copy the application's requirements.txt and run pip to install all
        # dependencies into the virtualenv.
        ADD requirements.txt /app/requirements.txt
        ADD my-gcp-key.json /app/my-gcp-key.json
        ENV GOOGLE_APPLICATION_CREDENTIALS /app/my-gcp-key.json
        RUN pip install -r /app/requirements.txt

        # Add the application source code.
        ADD . /app

        # Run a WSGI server to serve the application. waitress must be declared as
        # a dependency in requirements.txt.
        RUN pip install -e .

        CMD pserve development.ini

>     Replace ``my-gcp-key.json`` filename with the json file you were provided when you created the Service Account.

Reviewer: ```suggestion
    Replace ``my-gcp-key.json`` filename with the JSON file you were provided when you created the Service Account.
```

    This allows you to communicate with Datastore within every request.

Now let's assume you have the following model defined somewhere in your code that relates to a Datastore "kind":

.. code-block:: python

    from google.cloud import ndb

    class Accounts(ndb.Model):

        email = ndb.StringProperty()
        password = ndb.StringProperty()

        def __init__(self, **kwds):
            super(Accounts, self).__init__(**kwds)

You could then query this model within any handler/endpoint like so:

.. code-block:: python

     Accounts.query().filter(Accounts.email == user_email).get()

Running locally
---------------

Unlike App Engine's Standard environment, we're running Pyramid in a pretty typical fashion. You can run this locally on your machine using the same line in the ``dockerfile`` we created earlier as ``pserve development.ini``, or you can run in a Docker container using the same ``dockerfile`` that Flexible will be using. No changes need to be made there. This is useful for debugging any issues you may run in to under Flexible, without needing to deploy to it.

Deploying
---------

> Using the Google Cloud SDK, deploying is pretty straight-forward.

Reviewer: ```suggestion
Using the Google Cloud SDK, deploying is pretty straightforward.
```

.. code-block:: bash

    $ gcloud app deploy app.yaml --version my-version --project my-gcp-project

Replace ``my-version`` with some kind of identifier so you know what code is deployed. This can pretty much be anything.

> Replace ``my-gcp-project`` App Engine application's ID.

Reviewer: ```suggestion
Replace ``my-gcp-project`` with your App Engine application's ID.
```

    This allows you to communicate with Datastore within every request.

#.  ``production.ini``

    Your :app:`Pyramid` application should already contain both a ``development.ini`` and a ``production.ini``.
    For App Engine to communicate with your application, it will need to be listening on port 8080.
    Assuming you are using the Waitress WSGI server, modify the ``listen`` variable within the ``server:main`` block.

    .. code-block:: ini

>         listen = 0.0.0.0:8080

Reviewer: I'm not sure about using 0.0.0.0. For production we typically use *. See https://github.com/Pylons/pyramid-cookiecutter-starter/blob/latest/%7B%7Bcookiecutter.repo_name%7D%7D/production.ini#L42

Author: Let me try this in Flexible, certainly would be good to keep consistent.

Author: Yep, this works. Updated the file
